core/setup/kubernetes.go

```diff
@@ -47,47 +52,71 @@ func kubernetesParse(c *Controller) (kubernetes.Kubernetes, error) {
+//      Inflight:   &singleflight.Group{},
         APIConn:    nil,
-	}
-	var (
-		endpoints     = []string{defaultK8sEndpoint}
-	)
-	for c.Next() {
-		if c.Val() == "kubernetes" {
-			k8s.Zones = c.RemainingArgs()
-			if len(k8s.Zones) == 0 {
-				k8s.Zones = c.ServerBlockHosts
-			}
-			middleware.Zones(k8s.Zones).FullyQualify()
-			if c.NextBlock() {
-				// TODO(miek): 2 switches?
-				switch c.Val() {
-				case "endpoint":
-					args := c.RemainingArgs()
-					if len(args) == 0 {
-						return kubernetes.Kubernetes{}, c.ArgErr()
-					}
-					endpoints = args
+        NameTemplate: nil,
+		Namespaces: nil,
+    }
+    var (
+        endpoints = []string{defaultK8sEndpoint}
+        template = defaultNameTemplate
+		namespaces = []string{}
+    )
+
+    k8s.NameTemplate = new(nametemplate.NameTemplate)
+    k8s.NameTemplate.SetTemplate(template)
+
+    for c.Next() {
+        if c.Val() == "kubernetes" {
+            zones := c.RemainingArgs()
+
+            if len(zones) == 0 {
+                k8s.Zones = c.ServerBlockHosts
+            } else {
+                // Normalize requested zones
+                k8s.Zones = kubernetes.NormalizeZoneList(zones)
+            }
+
+            middleware.Zones(k8s.Zones).FullyQualify()
+            if c.NextBlock() {
+                // TODO(miek): 2 switches?
+                switch c.Val() {
+                case "endpoint":
+                    args := c.RemainingArgs()
+                    if len(args) == 0 {
+                        return kubernetes.Kubernetes{}, c.ArgErr()
+                    }
+                    endpoints = args
                     k8s.APIConn = k8sc.NewK8sConnector(endpoints[0])
-				}
-				for c.Next() {
-					switch c.Val() {
-					case "endpoint":
+                    fmt.Printf("[debug] Got k8s API connector: %v\n", k8s.APIConn)
+                }
+                for c.Next() {
+                    switch c.Val() {
+                    case "template":
+                        args := c.RemainingArgs()
+                        if len(args) == 0 {
+                            return kubernetes.Kubernetes{}, c.ArgErr()
+                        }
+                        template = strings.Join(args, "")
+                        err := k8s.NameTemplate.SetTemplate(template)
+						if err != nil {
+                            return kubernetes.Kubernetes{}, err
+						}
+					case "namespaces":
 						args := c.RemainingArgs()
 						if len(args) == 0 {
-							return kubernetes.Kubernetes{}, c.ArgErr()
+                            return kubernetes.Kubernetes{}, c.ArgErr()
 						}
-						endpoints = args
-					}
-				}
-			}
-			return k8s, nil
-		}
-        fmt.Println("endpoints='%v'", endpoints)
-	}
-	return kubernetes.Kubernetes{}, nil
+						namespaces = args
+						k8s.Namespaces = &namespaces
+                    }
+                }
+            }
+            return k8s, nil
+        }
+    }
+    return kubernetes.Kubernetes{}, nil
```

Comment:
This mirrors the etcd implementation. I don't see how the empty default is usable.

Might be better to error, or return an instance with the default values set.